File: exercises/javascript/exercise2/passwordValidator.test.js
const PasswordValidator = require('./passwordValidator');

describe('PasswordValidator', () => {
  let validator;

  beforeEach(() => {
    validator = new PasswordValidator();
  });

  test('null or empty strings return false', () => {
    expect(validator.validate(null)).toBe(false);
    expect(validator.validate('')).toBe(false);
    expect(validator.validate('   ')).toBe(false);
  });

  test('less than eight characters returns false', () => {
    expect(validator.validate('Abc123!')).toBe(false);
  });

  test('missing uppercase character returns false', () => {
    expect(validator.validate('password123!')).toBe(false);
  });

  test('missing digit returns false', () => {
    expect(validator.validate('PasswordDef!')).toBe(false);
  });

  test('contains sequential digits returns false', () => {
    expect(validator.validate('PassWord123!')).toBe(false);
  });

  test('valid complex password returns true', () => {
    expect(validator.validate('Str0ng!Pass9')).toBe(true);
  });
});

File: java/src/test/java/workshop/exercise2/PasswordValidatorTest.java
package workshop.exercise2;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.NullAndEmptySource;
import org.junit.jupiter.params.provider.ValueSource;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class PasswordValidatorTest {
    private PasswordValidator validator;

    @BeforeEach
    void setUp() {
        validator = new DefaultPasswordValidator();
    }

    @ParameterizedTest
    @NullAndEmptySource
    @ValueSource(strings = {"   "})
    void validate_NullOrEmpty_ReturnsFalse(String password) {
        assertFalse(validator.validate(password));
    }

    @Test
    void validate_LessThanEightCharacters_ReturnsFalse() {
        assertFalse(validator.validate("Abc123!"));
    }

    @Test
    void validate_MissingUppercase_ReturnsFalse() {
        assertFalse(validator.validate("password123!"));
    }

    @Test
    void validate_MissingDigit_ReturnsFalse() {
        assertFalse(validator.validate("PasswordDef!"));
    }

    @Test
    void validate_ContainsSequentialDigits_ReturnsFalse() {
        assertFalse(validator.validate("PassWord123!"));
    }

    @Test
    void validate_ValidComplexPassword_ReturnsTrue() {
        assertTrue(validator.validate("Str0ng!Pass9"));
    }
}
